Changed welcome bot messages

codewithnick opened the pull request:
Fixes #30049 

This image corresponds to all changes made in the 3 commits
1) Initial message made more concise
2) Reply to undefined message changed
3) Bot will not reply to any future undefined message

![image](https://github.com/zulip/zulip/assets/53932029/cf2170e1-ee51-4da8-a33f-85fa12d913ac)

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
@codewithnick Thanks! Please go through our [guide on submitting a pull request](https://zulip.readthedocs.io/en/latest/contributing/reviewable-prs.html) to learn how to present your proposed changes to Zulip, so that your work can be reviewed.

You'll need to redo the commit message explanations to follow our guidelines, and do a screenshot showing how you've tested this as per the guidelines and self-review checklist, as well as a few more things you'll see when reading that. 

Once you've done that, please ask for a review from @prakhar1144.

codewithnick commented:
![image](https://github.com/zulip/zulip/assets/53932029/40e92a95-ecae-401a-a2e5-b40f726c6d84)
this is the new message sent by the bot 

@prakhar1144 can you please check if I have done everything right this time, or missed something , before I move on to making other commits related to this issue

prakhar1144 commented:
Thanks. You need to read https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages
and re-write commit message accordingly.

codewithnick commented:
> Thanks. You need to read https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages and re-write commit message accordingly.
@prakhar1144 sorry to tag you again I am still trying to wrap my head around writing a good commit message

codewithnick commented:
Hello @prakhar1144 can you please review

prakhar1144 commented:
@codewithnick It's better. But still needs improvement like "Fixes..." should be on the last line. Please read the full page of the doc we linked above.

Also, go ahead and complete the remaining parts of the PR.

codewithnick commented:
task 2:

old message was
![image](https://github.com/zulip/zulip/assets/53932029/f807e6b6-97cb-4c92-aeb5-77341c37da41)

new message is
![image](https://github.com/zulip/zulip/assets/53932029/651e2b1e-6062-48ce-bfc9-46dbbca9d315)

partially fixes #30049

codewithnick commented:
@prakhar1144 please review , I think the 3rd task might need a bit of discussion before proceeding

codewithnick commented:
hi @prakhar1144 can you please review

alya commented:
Thanks! Please complete the self-review checklist (as applicable), and post a comment to request a review when ready.

You will also need to add your screenshots to the PR description, rather than just having them in a comment.

codewithnick commented:
ready for review @prakhar1144

prakhar1144 commented on zerver/lib/onboarding.py:
A better commit message would be
```
onboarding: Remove part of welcome bot initial direct message.

Earlier, the initial direct message sent by welcome bot suggested
a few bot commands to try. They don't seem super helpful at this
point.

To reduce the emphasis on commands, this commit updates the
message to drop that part.

Fixes part of #30049.
```

prakhar1144 replied:
Note that the doc mentions to include in your commit message: "why this commit improves the project. "

Also, a "." was missing in the "Fixes ..." line.

codewithnick replied:
Thanks! for the review , I'll change it

codewithnick commented:
@prakhar1144 Thanks for the review ,can you please check

codewithnick commented:
Thanks @alya can you please check this

alya commented:
@prakhar1144 When marking a PR for integration review, it's helpful to leave a brief comment (including a note about manual testing), and to remove the "maintainer review" label.

alya commented:
I don't really think this partial implementation of the issue is mergeable from a user experience perspective. @codewithnick , do you intend to finish implementing the issue? Otherwise, it might make sense for @prakhar1144 to take over from here.

codewithnick commented:
> I don't really think this partial implementation of the issue is mergeable from a user experience perspective. @codewithnick , do you intend to finish implementing the issue? Otherwise, it might make sense for @prakhar1144 to take over from here.

I'd like to finish it ,but need some help on the 3rd part so yes after a bit of discussion on the #backend thread

codewithnick commented:
@prakhar1144 can you please review the third commit, Thanks

prakhar1144 commented:
See #20460 for the design discussions for how it should work. It has some prior work.

The emoji should be added to the messages user sends + The "I'm happy to help you out" message should be sent before the bot reply.... and so on.

If possible you can resolve merge conflicts on #20460 (if it's not a lot of conflicts) as it already has some code reviews etc.

codewithnick commented:
> See #20460 for the design discussions for how it should work. It has some prior work.
> 
> The emoji should be added to the messages user sends + The "I'm happy to help you out" message should be sent before the bot reply.... and so on.
> 
> If possible you can resolve merge conflicts on #20460 (if it's not a lot of conflicts) as it already has some code reviews etc.

Thanks @prakhar1144 , this is what @alya suggested at the end of the conversation
https://github.com/zulip/zulip/pull/20460#issuecomment-1006172986

>  the first message to Welcome Bot (no matter what it is) the bot:
> 
> Reacts with a 🎉
> Sends an additional message that says: "I'm happy to help you out! 🎉 ", before sending the appropriate reply.

Sounds good ,will reimplement that 

@alya is it a good idea to randomly choose a confusion emote and send it if bot does not understand the message user sends after first time , I have implemented this in most recent commit, or should we do something else, like react with a ❓on the user message, please do suggest

alya commented:
> @alya is it a good idea to randomly choose a confusion emote and send it if bot does not understand the message user sends after first time [reference 3rd part of https://github.com/zulip/zulip/issues/20023 ], I have implemented this in most recent commit, or should we do something else, like react with a ❓on the user message, please do suggest

I'm currently thinking we want to keep it positive, and not have the Welcome Bot express confusion for messages it doesn't understand. We're giving the user an opportunity to test out Zulip in this DM thread. No reactions for the messages the bot doesn't understand is probably a fine starting point.

codewithnick commented:
Thanks, that is what I was thinking too, will implement this.

codewithnick commented:
@prakhar1144  @alya done

codewithnick commented:
hi @prakhar1144 can you please review this.

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -244,6 +244,31 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+            sender_id=send_request.message.sender.id,
+            recipient_id=send_request.message.recipient.id,
+        ).count()
+        == 1
```
Verify that it uses `zerver_message_realm_sender_recipient` index using `explain analyze` in dbshell (See Message model for the index). 

Update the comment in the Message model for where the index is used & add a "Uses index 'zerver_message_realm_sender_recipient'" comment here. -- It helps to be cautious while changing query or removing index in the future.

codewithnick replied:
Thanks, how can I verify this, I did not understand.

prakhar1144 replied:
You can use `python manage.py dbshell`

Then [explain analyze](https://www.postgresql.org/docs/current/sql-explain.html) to get an idea of the query plan.

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -244,6 +244,31 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+            emoji_code=tada_emoji_code,
+            reaction_type=tada_reaction_type,
+            emoji_name=emoji_name,
+        )
```
Can be simplified to:
```py
        emoji_data = get_emoji_data(realm_id=realm_id, emoji_name="tada")
        do_add_reaction(
            welcome_bot,
            send_request.message,
            emoji_data.emoji_name,
            emoji_data.emoji_code,
            emoji_data.reaction_type,
)
````

prakhar1144 commented on zerver/tests/test_tutorial.py:
```diff
@@ -186,3 +186,43 @@ def test_no_response_to_group_pm(self) -> None:
         self.send_personal_message(user1, bot, content)
         self.assertEqual(message_stream_count(user1), user1_messages + 2)
+
+    def test_response_to_pm_for_first_message(self) -> None:
```
We can keep all the checks in a single test. 

send first message -> verify reaction -> verify response -> send second message -> verify response

prakhar1144 reviewed (commented):
The PR doesn't implement the third point completely -- "Do not send a reply to any future message that Welcome Bot doesn't understand."

---

@alya is it fine to react on 

<details><summary>Screenshot</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/abe133a5-86df-4980-a152-ad7b533fffc4">
</img>
</details> 

confirming as you mentioned "_No reactions for the messages the bot doesn't understand is probably a fine starting point._" in the previous comment and

```
the first message to Welcome Bot (no matter what it is) the bot:

Reacts with a 🎉
```
 in https://github.com/zulip/zulip/pull/20460#issuecomment-1006172986
 
 I think it's fine to react and say "I'm happy to help you out!🎉 " because we're actually helping them out by suggesting bot commands (app, stream, topic, etc) to use.

alya commented:
> I think it's fine to react and say "I'm happy to help you out!🎉 " because we're actually helping them out by suggesting bot commands (app, stream, topic, etc) to use.

Sure, sounds good to me.

alya commented:
I'm confused: where does this come from?

> "I'm happy to help you out!🎉 

I don't see that in the spec.

alya commented:
Let's not react to random text. We don't really want to 🎉 in response to, "I'm confused." or something.

prakhar1144 commented:
> I'm confused: where does this come from?
> 
> > "I'm happy to help you out!🎉 
> 
> I don't see that in the spec.

This comment: https://github.com/zulip/zulip/pull/20460#issuecomment-1006172986

alya commented:
Ah, please drop both that reaction and the text. https://github.com/zulip/zulip/issues/30049 is the up-to-date version of our plans here.

codewithnick commented on zerver/lib/onboarding.py:
```diff
@@ -244,6 +244,31 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
     human_response_lower = send_request.message.content.lower()
     content = select_welcome_bot_response(human_response_lower)
+    human_recipient_id = send_request.message.sender.recipient_id
+    assert human_recipient_id is not None
```
I think i remember adding that line because sonarlint had started to complain

codewithnick commented:
Reopening the pr due to merge conflicts

codewithnick commented:
Hi @alya i'm not completely sure if this is what was expected in the 3rd task of #30049 
![image](https://github.com/zulip/zulip/assets/53932029/5154f8a5-7410-4fbc-95d6-3d83c94faad9)
The bot will not reply to any message it does not understand unless it is the first message by the user.
Please advice if something else was expected or I misunderstood.

codewithnick commented:
Thanks for your patience @prakhar1144 ,I have updated the pr to match the issue, can you please do a review again.

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -191,7 +187,23 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+            sender_id=welcome_bot.id,
+            recipient__type_id=human_user_id,
+            content__icontains=bot_commands(),
+        ).exists()
```
This is the query plan:

```
zulip=> explain analyze SELECT 1 AS "a" FROM "zerver_message" INNER JOIN "zerver_recipient" ON ("zerver_message"."recipient_id" = "zerver_recipient"."id") WHERE (UPPER("zerver_message"."content"::text) LIKE UPPER('%`apps`, `profile`, `theme`, `channels`, `topics`, `message formatting`, `keyboard shortcuts`, `help`.%') AND "zerver_message"."realm_id" = 2 AND "zerver_recipient"."type_id" = 10 AND "zerver_message"."sender_id" = 5) LIMIT 1;
                                                                            QUERY PLAN                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.14..18.00 rows=1 width=4) (actual time=0.031..0.031 rows=0 loops=1)
   ->  Nested Loop  (cost=0.14..18.00 rows=1 width=4) (actual time=0.030..0.030 rows=0 loops=1)
         ->  Seq Scan on zerver_recipient  (cost=0.00..1.64 rows=2 width=4) (actual time=0.006..0.010 rows=2 loops=1)
               Filter: (type_id = 10)
               Rows Removed by Filter: 78
         ->  Index Scan using zerver_message_realm_sender_recipient on zerver_message  (cost=0.14..8.17 rows=1 width=4) (actual time=0.009..0.009 rows=0 loops=2)
               Index Cond: ((realm_id = 2) AND (sender_id = 5) AND (recipient_id = zerver_recipient.id))
               Filter: (upper(content) ~~ '%`APPS`, `PROFILE`, `THEME`, `CHANNELS`, `TOPICS`, `MESSAGE FORMATTING`, `KEYBOARD SHORTCUTS`, `HELP`.%'::text)
 Planning Time: 23.432 ms
 Execution Time: 0.081 ms
(10 rows)

```

@timabbott do you think sequential scan on Recipient table is fine ? -- I expect the table to be huge...

codewithnick replied:
I dont have any Idea how to do this can you please refer me to some docs, Thanks again

prakhar1144 replied:
See: https://github.com/zulip/zulip/pull/30278#discussion_r1659013545

codewithnick replied:
Thank you so much

prakhar1144 commented:
Works fine in manual testing. Suggested a few improvements above. 

You can address the first two review comments, we'll wait for Tim to answer the third comment.

codewithnick commented on zerver/lib/onboarding.py:
```diff
@@ -187,7 +187,23 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+            content__icontains=bot_commands(),
+        ).exists()
+        and bot_commands() in content
+    ):
```
Thanks prakhar ,for my future reference can you please explain the second part more, how is the variable helping here?

prakhar1144 replied:
```
            content__icontains=bot_commands(),
        ).exists()
        and bot_commands() in content
```
The current code calls `bot_commands()` twice. This function does some string concatenation, joins.. etc Using the variable we can avoid those string operations happening twice. Computationally doesn't affect much, but that's a small improvement we can make.

codewithnick commented:
# After review changes
[x] added comments to test_tutorial.py
[x] added a test case to check if bot is replying to defined messages after not replying to undefined messages
[x] filtering based on `recipient=human_user.recipient` instead of `recipient__type_id=human_user_id`

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -187,7 +187,23 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+        and Message.objects.filter(
+            realm_id=realm_id,
+            sender_id=welcome_bot.id,
+            recipient=human_user.recipient,
```
Use `recipient_id=human_user.recipient_id,` here.

It'll save a db query to get the recipient row. We only need ID of the recipient here.

codewithnick replied:
oh yeah right

prakhar1144 reviewed (commented):
I think you missed this comment: https://github.com/zulip/zulip/pull/30278#discussion_r1650553635

codewithnick commented:
# after review changes

![image](https://github.com/zulip/zulip/assets/53932029/6f8ccd70-6a13-42bc-ab4b-6ec168f9856e)
filtering based on `recipient_id` instead of `recipient`

added a check for `human_user.recipeint_id` is not `NONE`
![image](https://github.com/zulip/zulip/assets/53932029/025a446e-e04a-47b6-886e-cbbda221a11e)

prakhar1144 commented:
```
human_user_recipient_id = send_request.message.sender
assert human_user_recipient_id is not None
```

codewithnick commented:
are you sure thats correct `message.sender` is of type `UserProfile` and not of type `Recipient`

prakhar1144 commented:
Yeah, that's what I meant. Sorry for the typo.

```
human_user_recipient_id = send_request.message.sender.recipient_id
assert human_user_recipient_id is not None
```

codewithnick commented:
# After review changes 

Changed to `human_user_recipient_id`

prakhar1144 commented:
@codewithnick I think you missed https://github.com/zulip/zulip/pull/30278#pullrequestreview-2150497169 again :)

codewithnick commented:
@prakhar1144 added comments

prakhar1144 commented on zerver/models/messages.py:
```diff
@@ -186,6 +186,8 @@ class Meta:
                 # sender_id) can be used for scrubbing users and/or
                 # deleting users' messages.
+                # Also used to check welcome bot messages
+                # while onboarding
```
This is not restricted to onboarding. Maybe simply: "Also used in 'send_welcome_bot_response'."

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -187,7 +187,25 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+    ):
+        # If the bot has already sent bot commands to this user and
+        # If the bot does not understand the current message sent by this user then
+        # Do not send any message
```
nit: The first letter of If (second line) and Do (third line) won't be in uppercase.

prakhar1144 commented on zerver/lib/onboarding.py:
```diff
@@ -187,7 +187,25 @@ def send_welcome_bot_response(send_request: SendMessageRequest) -> None:
+            recipient_id=human_user_recipient_id,
+            content__icontains=commands,
+        ).exists()
+        # Uses index 'zerver_message_realm_sender_recipient'
```
I checked after populating database with 3000 messages.. It uses 
```py
                                                                                             QUERY PLAN                                                                                             
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.28..4.31 rows=1 width=4) (actual time=0.068..0.069 rows=0 loops=1)
   ->  Index Scan using zerver_message_realm_id_849a39c8 on zerver_message  (cost=0.28..4.31 rows=1 width=4) (actual time=0.065..0.066 rows=0 loops=1)
         Index Cond: (realm_id = 2)
         Filter: ((recipient_id = 119) AND (sender_id = 5) AND (upper(content) ~~ '%`APPS`, `PROFILE`, `THEME`, `CHANNELS`, `TOPICS`, `MESSAGE FORMATTING`, `KEYBOARD SHORTCUTS`, `HELP`.%'::text))
 Planning Time: 0.821 ms
 Execution Time: 0.113 ms
(6 rows)
```

Probably @timabbott should run this query for CZO and then we can update the index getting used accordingly.

codewithnick replied:
is there a er diagram/schema for zulip?

prakhar1144 replied:
I don't think so.

Can you post the above index related query in #backend thread on CZO for faster response ?

prakhar1144 replied:
We use `zerver_message_realm_sender_recipient` as [confirmed by Tim on CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/query.20message.20sent.20by.20welcome.20bot/near/1851702)

prakhar1144 commented:
Looks good to me, now.

Works in manual testing as expected.

Tagging it for integration review.

alya commented:
Works for me in manual testing!

alya commented:
I added a commit to update Welcome Bot's responses to commands, which had gotten stale.
: 
<details>
<summary>
Screenshot of changed response
</summary>

![Screenshot 2024-07-08 at 23 59 37](https://github.com/zulip/zulip/assets/2090066/f104d618-ce98-48c5-9701-97f3091af571)
</details>



@prakhar1144 if you are up for fixing any tests I might have broken, that would be helpful. :)

timabbott commented:
This was merged via #30789, thanks for the contribution @codewithnick!
